src/Microsoft.SqlTools.ServiceLayer/BatchParser/BatchParserWrapper.cs

```diff
@@ -0,0 +1,423 @@
+        /// <param name="positions"></param>
+        /// <returns></returns>
+        private static IList<int> ConvertToOffsetSortedInput(string content, IList<System.Tuple<int, int>> positions)
+        {
+            IList<int/*offset*/> offsetList = new List<int>();
+
+            if (!string.IsNullOrEmpty(content) && (positions.Count > 0))
+            {
+                // Get char array to find \r and \n
+                using (StringReader reader = new StringReader(content))
+                {
+                    int lineStartOffset = 0;
+                    int count = positions.Count;
+
+                    // Calculate the first offset
+                    // positions is 1-based
+                    int lineDifference = positions[0].Item1;
+                    lineStartOffset = ReadLines(reader, lineStartOffset, lineDifference);
+                    // offset is 0-based
+                    int offset = lineStartOffset + positions[0].Item2;
+                    offsetList.Add(offset);
+
+                    // Calculate the rest offsets
+                    for (int index = 1; index < count; index++)
+                    {
+                        lineDifference = positions[index].Item1 - positions[index - 1].Item1;
+                        lineStartOffset = ReadLines(reader, lineStartOffset, lineDifference);
+                        offset = lineStartOffset + positions[index].Item2;
+                        offsetList.Add(offset);
+                    }
+                }
+            }
+            return offsetList;
+        }
+
+        /// <summar>
+        /// Helper method used to get the end column
+        /// <param name="reader"></param>
+        /// <param name="endLine"></param>
+        /// <returns></returns>
+        private static int GetEndColumn(StringReader reader, int endLine)
+        {
+            string endLineString = null;
+            int endColumn = 0;
+            int ch;
+            for (int i = 0; i < endLine; i++) {
+                endLineString = reader.ReadLine();
+            }
+            StringReader endLineReader = new StringReader(endLineString);
+            while (true) {
+                ch = endLineReader.Read();
+                if (ch == -1) 
+                {
+                    return endColumn;
+                }
+                endColumn++;
+            }
+        }
+
+        /// <summary>
+        /// Helper method used to get the end line and column for the last batch
+        /// </summary>
+        /// <param name="reader"></param>
+        /// <returns></returns>
+        private static System.Tuple<int /*endLine */, int /*endColumn */> GetLastEndLineColumn(StringReader reader)
+        {
+            int endLine = 0;
+            int endColumn = 0;
+            string nextLine;
+            string lastLine = null; // get the last line
+            int ch;
+
+            while ((nextLine = reader.ReadLine()) != null)
+            {
+                endLine++;
```

Comment:
The code might be misleading, but essentially it goes to the last line of the file regardless. I do agree that starting endLine from the previous endLine would make it more efficient.